ps3.py

```diff
@@ -0,0 +1,374 @@
+    """
+    word = word.lower()
+    hand_modified = hand.copy()
+    for letters in word:
+        if letters in hand_modified.keys():
+            hand_modified[letters] -= 1
+    list_keys = list(hand_modified.keys())
+    for keys in list_keys:
+        if hand_modified[keys] <= 0:
+            hand_modified.pop(keys)
+    return hand_modified
+
+
+def is_valid_word(word, hand, word_list):
+    """
+    Returns True if word is in the word_list and is entirely
+    composed of letters in the hand. Otherwise, returns False.
+    Does not mutate hand or word_list.
+
+    word: string
+    hand: dictionary (string -> int)
+    word_list: list of lowercase strings
+    returns: boolean
+    """
+    word = word.lower()
+    for letters in word:
+        if letters not in hand.keys() or word.count(letters) > hand[letters]:
+            return False
+
+    if '*' in word:
+        list_split = word.split('*')
+        list_possible_words = [str(list_split[0] + i + list_split[1]) for i in VOWELS]
+        for words in list_possible_words:
+            if words in word_list:
+                return True
+        return False
+
+    if word not in word_list:
+        return False
+    return True
+
+
+def calculate_handlen(hand):
+    """
+    Returns the length (number of letters) in the current hand.
+
+    hand: dictionary (string-> int)
+    returns: integer
+    """
+    len_hand = 0
+    for keys in hand:
+        len_hand += hand[keys]
+    return len_hand
+
+
+def play_hand(hand, word_list):
+    """
+    Allows the user to play the given hand, as follows:
+
+    * The hand is displayed.
+
+    * The user may input a word.
+
+    * When any word is entered (valid or invalid), it uses up letters
+      from the hand.
+
+    * An invalid word is rejected, and a message is displayed asking
+      the user to choose another word.
+
+    * After every valid word: the score for that word is displayed,
+      the remaining letters in the hand are displayed, and the user
+      is asked to input another word.
+
+    * The sum of the word scores is displayed when the hand finishes.
+
+    * The hand finishes when there are no more unused letters.
+      The user can also finish playing the hand by inputing two
+      exclamation points (the string '!!') instead of a word.
+
+      hand: dictionary (string -> int)
+      word_list: list of lowercase strings
+      returns: the total score for the hand
+
+    """
+    # Keep track of the total score
+    # As long as there are still letters left in the hand:
+    total_score = 0
+    while True:
+        if calculate_handlen(hand) == 0:
+            print("Ran out of letters")
+            print("Total score for this hand: " + str(total_score))
+            break
+        # Display the hand
+        print("Current Hand:", end=' ')
+        display_hand(hand)
+        # Ask user for input
+        word = input("Enter word, or “!!” to indicate that you are finished: ")
+        if word == '!!':
```

Comment:
What do you think about constants for wildcard and '!!'?

Reply:
+